src/sync/arc.rs

```diff
@@ -5,36 +5,72 @@ use std::{mem, ops};
 
 /// Mock implementation of `std::sync::Arc`.
 #[derive(Debug)]
-pub struct Arc<T> {
-    inner: std::sync::Arc<Inner<T>>,
-}
-
-#[derive(Debug)]
-#[repr(C)]
-struct Inner<T> {
-    // This must be the first field to make into_raw / from_raw work
-    value: T,
-
-    obj: rt::Arc,
+pub struct Arc<T: ?Sized> {
+    obj: std::sync::Arc<rt::Arc>,
+    inner: std::sync::Arc<T>,
 }
 
 impl<T> Arc<T> {
     /// Constructs a new `Arc<T>`.
     #[track_caller]
     pub fn new(value: T) -> Arc<T> {
-        let inner = std::sync::Arc::new(Inner {
-            value,
-            obj: rt::Arc::new(location!()),
-        });
+        let std = std::sync::Arc::new(value);
 
-        Arc { inner }
+        Arc::from_std(std)
     }
 
     /// Constructs a new `Pin<Arc<T>>`.
     pub fn pin(data: T) -> Pin<Arc<T>> {
         unsafe { Pin::new_unchecked(Arc::new(data)) }
     }
 
+    /// Returns the inner value, if the `Arc` has exactly one strong reference.
+    pub fn try_unwrap(_this: Arc<T>) -> Result<T, Arc<T>> {
+        unimplemented!();
+    }
+}
+
+impl<T: ?Sized> Arc<T> {
+    /// Converts `std::sync::Arc` to `loom::sync::Arc`
+    ///
+    /// This is needed to create `Arc<T>` where `T: !Sized`
```

Comment:
`loom::sync::Arc::new` accepts the argument by value (just as `std::sync::Arc::new`) and so it can't accept `?Sized` types.

The same can be said about `std::sync::Arc::new`, it too can only be called with `T: Sized`. The magic, that allows creating `std::sync::Arc` with `!Sized` types is "unsize-coercion" via [`CoerceUnsized`](https://doc.rust-lang.org/std/sync/struct.Arc.html#impl-CoerceUnsized%3CArc%3CU%3E%3E).

Reply:
ah, yeah, that makes sense. thanks!